text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,239 @@
+
+There is nothing preventing private `extern existential type`, or a `extern extern existential type` in the same module as its `extern existential type`.
+Both these situations make the feature useless and could be linted, but are well defined from the rules above.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The fact that not all crates can compose with this, due to duplicate or missing definitions per declaration, is not very nice.
+However, this is exactly like "lang items" and the annotations that exist for this purpose today, so it is nothing worse than what's currently about to be stabilized.
+There is no natural orphan rules for this feature (or alternatively, regular `existential type` can be seen as this with the orphan rule that it must be defined in the same module), so this is expected.
+See the first alternative for how we can use Cargo to ameliorate this.
+
+Niko Matsakis has expressed concerns about this being abused because singletons are bad.
+Singletons are indeed bad, but the connection between existential types and singletons is not obvious at first sight (imagine if we had deferred definition mechanism with `static`s directly), which hopefully will make this be sufficiently difficult to abuse.
+Even if we deem this very toxic, better all the use cases I listed above be white-listed and use same mechanism used for consistency (and one that is cost-free at run time), than use a bunch of separate solutions.
+Also, by forcing the use of a trait in the bounds of the `extern existential type`, we hopefully nudge the user in the direction of providing a non-singleton-based way of accomplishing the same task (e.g. local allocators in addition to the global allocator).
+
+Stabilization of many annotations and APIs called out in the [motivation](#motivation) section is imminent, and yes this would delay that a bit if we opted to do this and then rewrite those APIs to use it.
+
+As per the [prior art](#prior-art) section, something like Haskell's backpack is wholly superior.
+But as stabilization of the status quo is imminent, I wanted to pick something easier to implement and closer to existing rust features mentally/pedagogically.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+- We can additionally mandate that `Cargo.toml` include all `extern existential type` declarations and definitions, and Cargo reject any build plan where they don't match 1-1.
+  This ameliorates the crate composition issue in practice for the vast majority of users using Cargo (even just `Cargo.toml`s).
+
+- Of course, we can always do nothing and just keep the grab bag of ad-hoc solutions we have today, and leave log with just a imperative dynamic solution.
+
+- We could continue special-casing and white-listing in the compiler the use-cases I give in the motivation, but at least use the same sort of annotation for all of them for consistency.
+  But that still requires leaving out `log`, or special casing it for the first time.
```

Comment:
I don't think we'd want to use this for log at all. Runtime configuration of loggers is pervasive.

Reply:
Who does that run-time configuration? The libraries logging or the end application?